src/cm/communication_manager.rs

```diff
@@ -56,55 +229,356 @@ impl EventChannel {
+        }
+
+        Ok(())
+    }
+
+    pub fn resolve_addr(
+        &self, src_addr: Option<SocketAddr>, dst_addr: SocketAddr, timeout: Duration,
+    ) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let timeout_ms: i32 = timeout.as_millis().try_into().unwrap();
+
+        let ret = unsafe {
+            rdma_resolve_addr(
+                cm_id.as_ptr(),
+                match src_addr {
+                    Some(addr) => OsSocketAddr::from(addr).as_mut_ptr(),
+                    None => null_mut(),
+                },
+                OsSocketAddr::from(dst_addr).as_mut_ptr(),
+                timeout_ms,
+            )
+        };
 
         if ret < 0 {
-            return Err(format!("Failed to bind addr {addr:?}, returned {ret}"));
+            return Err(format!("Failed to resolve address {ret}"));
         }
 
         Ok(())
     }
 
-    pub fn listen(&mut self, backlog: i32) -> Result<(), String> {
-        let ret = unsafe { rdma_listen(self.cm_id.as_mut(), backlog) };
+    pub fn resolve_route(&self, timeout: Duration) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let timeout_ms: i32 = timeout.as_millis().try_into().unwrap();
+
+        let ret = unsafe { rdma_resolve_route(cm_id.as_ptr(), timeout_ms) };
+
+        if ret < 0 {
+            return Err(format!("Failed to resolve route {ret}"));
+        }
+
+        Ok(())
+    }
+
+    pub fn listen(&self, backlog: i32) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let ret = unsafe { rdma_listen(cm_id.as_ptr(), backlog) };
 
         if ret < 0 {
             return Err(format!("Failed to listen {ret}"));
         }
 
         Ok(())
     }
+
+    pub fn get_device_context(&self) -> Option<Arc<DeviceContext>> {
+        let cm_id = self.cm_id;
+
+        unsafe {
+            let mut guard = DEVICE_LISTS.lock().unwrap();
+            let device_ctx = guard
+                .entry((*cm_id.as_ptr()).verbs as usize)
+                .or_insert(Arc::new(DeviceContext {
+                    context: (*cm_id.as_ptr()).verbs,
+                }));
+
+            Some(device_ctx.clone())
+        }
+    }
+
+    pub fn connect(&self, mut conn_param: ConnectionParameter) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let ret = unsafe { rdma_connect(cm_id.as_ptr(), &mut conn_param.0) };
+
+        if ret < 0 {
+            return Err(format!("Failed to connect {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn disconnect(&self) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let ret = unsafe { rdma_disconnect(cm_id.as_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to disconnect {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn accept(&self, conn_param: Option<ConnectionParameter>) -> Result<(), String> {
```

Comment:
Since we are not willing to support binding QP to the cm_id, maybe we can define the `conn_param` as `ConnectionParameter` rather than `Option<ConnectionParameter>`.